framework/Optimizers/GradientBasedOptimizer.py

```diff
@@ -291,32 +401,307 @@ def checkConvergence(self):
     """
     convergence = True
     for traj in self.optTraj:
-      if self.convergeTraj[traj] == False:
+      if not self.convergeTraj[traj]:
         convergence = False
         break
     return convergence
 
-  def _updateConvergenceVector(self, traj, varsUpdate, currentLossVal):
+  def _checkModelFinish(self, traj, updateKey, evalID):
+    """
+      Determines if the Model has finished running an input and returned the output
+      @ In, traj, int, traj on which the input is being checked
+      @ In, updateKey, int, the id of variable update on which the input is being checked
+      @ In, evalID, int or string, indicating the id of the perturbation (int) or its a variable update (string 'v')
+      @ Out, _checkModelFinish, tuple(bool, int), (1,realization dictionary),
+            (indicating whether the Model has finished the evaluation over input identified by traj+updateKey+evalID, the index of the location of the input in dataobject)
+    """
+    if len(self.mdlEvalHist) == 0:
+      return (False,-1)
+    lookFor = '{}_{}_{}'.format(traj,updateKey,evalID)
+    index,match = self.mdlEvalHist.realization(matchDict = {'prefix':lookFor})
+    # if no match, return False
+    if match is None:
+      return False,-1
+    # otherwise, return index of match
+    return True, index
+
+  def _createEvaluationIdentifier(self,trajID,iterID,evalType):
```

Comment:
Right now we have no alternatives to the GradientBased for creating evaluation identifiers; I could create a dummy that sits in that space, but there are no methods in the base class that use the identifiers, either; it's a feature specific to GradientBased inheritors until we have another non-gradient type. Added the dummy for now.

Reply:
Paul, I know we have no alternatives to the GradientBased opts now, but new optimizers are coming and having a method that sets a standard does not hurt, also because, independently on the optimization strategy, identifiers need to be generated.

Reply:
Yeah, I get you.  I added it.